feat: 종목 상세페이지 기능 일부 연동

jihyeon-songb opened the pull request:
## ✨ 개요

<!-- 이번 PR에서 무엇을 했는지 간단히 설명 (ex: 회원가입 이메일 중복 검사 기능 추가) -->
종목 상세페이지 기능 일부 연동

## 💡 변경 내용

<!-- 주요 변경사항을 요약해서 bullet point로 작성해주세요. -->

- prettier 목록의 마지막 요소 뒤에 , 추가 옵션 삭제
- 등락값 계산 로직 공통 함수로 분리 및 RankingRow 리팩토링
- websocket 타입 분리 및 useStockRankingWebSocket 훅 이름 변경
- 종목 정보 api 추가
- 종목 일별 시세 api + 페이지네이션 추가
- 종목 최근 시세 api 함수 추가
- 종목 현재가 실시간 웹소켓 훅 추가
- 종목 아이콘 이미지, 한글 이름, 심볼, 현재가, 등락률 데이터 연동
- 종목 상세 차트 심볼 데이터 연동 및 차트 시간대 변경 기능 추가
- react-intersection-observer 라이브러리 설치
- 일별 시세 데이터 연동
- 무한스크롤 기능 추가
- TickerTapeWidget 푸터 추가
- ordersAtom 추가


## 📸 스크린샷 (선택 사항)

<!-- UI/UX 변경 사항이 있을 경우, 관련 스크린샷이나 GIF를 Before / After 캡처를 첨부해주세요. -->
<img width="1512" height="826" alt="image" src="https://github.com/user-attachments/assets/943c047c-1d7f-452a-a7f8-c30348424ff6" />


## 🧩 관련 이슈

<!-- PR과 관련된 이슈 번호를 링크하세요. -->

- Issue: #59 

## 🧐 리뷰어에게

<!-- 리뷰 시 중점적으로 확인할 부분이나 주의할 점을 작성해주세요 -->
종목 상세페이지 기능 일부 연동하였습니다. 
무한스크롤을 react-intersection-observer를 활용해 구현했는데, 데이터 페칭 및 페이지네이션 처리 방식이 적절한지 확인 부탁드립니다.

## ✅ Checklist

- [x]  커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x]  Self-Review를 완료했습니다.
- [x]  문서가 필요한 경우, 관련 문서를 업데이트했습니다.

jihyeon-songb committed:
chore: prettier 목록의 마지막 요소 뒤에 , 추가 옵션 삭제

jihyeon-songb committed:
refactor: 등락값 계산 로직 공통 함수로 분리 및 RankingRow 리팩토링

jihyeon-songb committed:
refactor: websocket 타입 분리 및 useStockRankingWebSocket 훅 이름 변경

jihyeon-songb committed:
feat: 종목 타이틀 데이터 연동 및 현재가 실시간 웹소켓 연결

- SymbolInfo type 정의
- fetchSymbolInfo api 함수 추가
- fetchSymbolTimeSeries api 함수 추가
- fetchRecentDailyPrice api 함수 추가
- 종목 현재가 실시간 웹소켓 훅 추가
- 종목 아이콘 이미지, 한글 이름, 심볼, 현재가, 등락률 데이터 연동

jihyeon-songb committed:
feat: 종목 상세 차트 심볼 데이터 연동 및 차트 시간대 변경 기능 추가

- 종목 상세 차트 데이터 연동
- 시간대 드롭 다운 ui 추가
- ordersAtom으로 symbol, chart timeframe 관리
- ChartTimeFrame 가능한 시간대 타입만 정의

jihyeon-songb committed:
chore: react-intersection-observer 라이브러리 설치

jihyeon-songb committed:
style: table ui css 수정

-

jihyeon-songb committed:
feat: 일별 시세 데이터 연동, 무한스크롤 기능 추가 및 TickerTapeWidget 푸터 추가

- 종목 상세페이지 부분 css 수정
- TickerTapeWidget 푸터 추가
- 일별 시세 데이터 연동 및 무한스크롤 기능 추가
- useSymbolTimeSeries.ts 추가
- useInfiniteScroll.ts 추가

pro-flab commented on src/types/websocket.ts:
```diff
@@ -0,0 +1,17 @@
+export const WS_URL = "wss://stream.data.alpaca.markets/v2/delayed_sip";
+
+export type RTRow = {
+  symbol: string;
+  last?: number;
+  size?: number;
+  ts?: string;
+};
+export type MSGRow = {
+  S: string;
+  T: string;
+  c: Array<string>;
+  i: number;
+  p: number;
+  s: number;
+  t: string;
+};
```
RTRow와 MSGRow가 타입만 봐서는 어떤걸 의미하는지, 그리고 내부 필드들도 구현을 보지 않으면 그 의미를 알기가 매우 어려워보입니다!

최대한 네이밍을 직관적으로 만들되, 그게 어려울 경우 주석을 달아주시면 좋겠습니다 🙇

pro-flab commented on src/pages/StockOrders.tsx:
```diff
@@ -2,36 +2,82 @@ import { useNavigate, useParams } from "react-router";
     type ?? "order",
   );
 
+  useEffect(() => {
+    if (!symbol) return;
+    setSymbolAtomValue(symbol);
+  }, [symbol, setSymbolAtomValue]);
+
+  const { data: symbolInfo } = useQuery({
+    queryKey: ["symbol-info", symbol],
+    queryFn: () => fetchSymbolInfo(symbol!),
+    enabled: !!symbol,
+  });
+
+  const { data: symbolPrice } = useQuery({
+    queryKey: ["symbol-price", symbol],
+    queryFn: () => fetchRecentDailyPrice(symbol!),
+    enabled: !!symbol,
+  });
+
```
[참고] 이렇게 해도 전혀 문제는 없지만, 보통 useEffect는 Hook들 중 맨 아래에 두어 Effect를 찾기 쉽게 하기 위한 컨벤션이 있습니다ㅎㅎ

pro-flab commented on src/pages/StockOrders.tsx:
```diff
@@ -2,36 +2,82 @@ import { useNavigate, useParams } from "react-router";
+  const { data: symbolInfo } = useQuery({
+    queryKey: ["symbol-info", symbol],
+    queryFn: () => fetchSymbolInfo(symbol!),
+    enabled: !!symbol,
+  });
+
+  const { data: symbolPrice } = useQuery({
+    queryKey: ["symbol-price", symbol],
+    queryFn: () => fetchRecentDailyPrice(symbol!),
+    enabled: !!symbol,
```
이렇게 ! 연산자를 이중으로 사용해서 값의 유무를 체크하는 것은 좋지 않은 코드라고 생각해요. 그 의미를 이해하는게 어렵기 때문인데요.

앞으로 이런 류의 코드를 쓸 때 아래와 같이 가져가보면 어떨까요?


1. `boolean`이 아닌 값에 대해
  - 값의 유무를 체크할 때 `symbol != null 혹은 symbol == null`
3. `boolean`인 값에 대해 
  - `false`인지 체크할 때, `!symbol`
  - `true`인지 체크할 때, 그냥 `symbol`
  - 만약 nullish한 값일 경우 null인지 체크할 때, `symbol != null`

pro-flab commented on src/pages/StockOrders.tsx:
```diff
@@ -2,36 +2,82 @@ import { useNavigate, useParams } from "react-router";
+  const prevClosePrice = timeSeriesItems[0]?.close;
+
+  const { rtSymbolPrice } = useStockDetailWebSocket(
+    symbol!,
```
symbol에 `!`(non-null assertion operator)를 사용하여 타입을 강제로 null이 아니게 했어요. 이것은 정말 극단적인 상황 외에는 절대 사용해서는 안되는 연산자인데요.

이런 경우 어떻게 해결할 수 있을지 멘토링 시간에 같이 얘기해봐요.

jihyeon-songb committed:
fix: 거래대금 Text 오류 수정

pro-flab commented on src/components/orders/StockTitle.tsx:
```diff
@@ -1,33 +1,53 @@
             className="rounded-xl"
-            src="/src/assets/profile.webp"
+            src={
+              symbolInfo.branding_icon_url +
+              `?apiKey=${import.meta.env.VITE_POLYGON_API_KEY}`
```
apiKey를 query string으로 집어넣는군요.. 정확히 어떤 것인지 멘토링 시간에 같이 얘기를 듣고싶네요 :)

pro-flab commented on src/hooks/useInfiniteScroll.ts:
```diff
@@ -0,0 +1,21 @@
+  hasNextPage: boolean | undefined,
+  isFetchingNextPage: boolean,
+) => {
+  const { ref: loadMoreRef, inView } = useInView({
+    threshold: 0,
+    triggerOnce: false,
+  });
```
이 useInView의 hook이 내부적으로 어떻게 동작하는지 알고 계신가요! 요 부분에 대해서 멘토링 시간에 얘기해봐요. 이 구현에 대해 자세히 알고 있는게 중요합니다 :)

pro-flab commented on src/hooks/useSymbolTimeSeries.ts:
```diff
@@ -0,0 +1,34 @@
+    queryFn: ({ pageParam }: { pageParam: Cursor | null }) =>
+      fetchSymbolTimeSeries(symbol, pageParam, pageSize),
+    getNextPageParam: (lastPage) => lastPage.nextCursor,
+    enabled: !!symbol, // symbol이 있을 때만 실행
```
symbol이 nullish한 값일 수도 있다는건데, 실제로 symbol은 그냥 `string`으로만 받고 있네요.

jihyeon-songb committed:
feat: 종목 상세 실시간 시세 기능 추가

- 실시간, 일별 시세 컴포넌트 세분화
- formatHHmmss utils 함수 추가
- 실시간 시세 atom(`realtimeSymbolPriceAtom`) 추가
- 종목 전일 종가 atom(`prevClosePriceAtom`) 추가
- StockPricePanelRow -> DailyPriceTableRow 컴포넌트 네이밍 변경
- !!symbol -> symbol != null로 조건식 명시적 변경
- useEffect 호출 위치 hook 맨 아래로 자리 이동
- RTRow, MSGRow -> RealtimeRow, SocketMsgRow 타입 네이밍 변경

jihyeon-songb committed:
refactor: 네이밍 정리 및 불필요 코드 제거

- MySelect → SelectTemplate 컴포넌트 파일명 변경
- 상수를 함수 바깥 스코프로 이동
- 불필요한 as 타입 단언 삭제
- 매개변수 타입이 동일한 경우 함수 단순화
- 불필요한 useEffect 제거
- 불필요한 useMemo 제거

jihyeon-songb committed:
chore: 실시간 시세 최신순으로 변경
